# Lab5 Równania nieliniowe
### Autor: Krzysztof Hardek

In [45]:
from mpmath import *

Funkcje Do testów

In [46]:
def f1(x):
    return mpf(cos(x)*cosh(x) - 1)
    
    
def f2(x):
    return mpf((1/x) - tan(x))
    
    
def f3(x):
    return mpf(2**(-x) + e**x + 2 * cos(x) - 6)   

Ich pochodne

In [47]:
def f1_d(x):
    return cos(x) * sinh(x) - sin(x) * cosh(x)


def f2_d(x):
    return (-1/(x**2)) - (1/(cos(x)**2))


def f3_d(x):
    return -ln(2) * 2**(-x) + e**x - 2 * sin(x)

## Zadanie 1 Metoda bisekcji
Funkcje realizująca bisekcje oraz wypisująca rezultat

In [48]:
def bisection(f, precision, a, b, error):
    mp.dps = precision
    x0 = mpf((a + b)/2)
    i = 0
    
    while abs(f(x0) - 0) > error:
        if f(a) * f(x0) < 0:
            b = x0
        else:
            a = x0
        i += 1
        x0 = mpf((a + b)/2)
        
    return [x0, i]


def print_result(f, x0, iterations):
    print(f'funkcja {f.__name__}, miejsce zerowe: {x0}, f(x0): {f(x0)}, liczba iteracji: {iterations}')     

Liczba iteracji potrzebnych do osiągnięcia dokładności $ 10^{-7} $ 

In [49]:
a = bisection(f1, 12, 3*pi/2, 2*pi, mpf(10)**(-7))
print_result(f1, a[0], a[1])
b = bisection(f2, 12, 0.0001, pi/2, mpf(10)**(-7))
print_result(f2, b[0], b[1])
c = bisection(f3, 12, 1, 3, mpf(10)**(-7))
print_result(f3, c[0], c[1])

funkcja f1, miejsce zerowe: 4.73004074596, f(x0): 6.33440322417e-8, liczba iteracji: 28
funkcja f2, miejsce zerowe: 0.860333566189, f(x0): 8.45204795041e-8, liczba iteracji: 20
funkcja f3, miejsce zerowe: 1.82938361168, f(x0): 3.99704731535e-8, liczba iteracji: 22


Liczba iteracji potrzebnych do osiągnięcia dokładności $ 10^{-15} $ 

In [50]:
a = bisection(f1, 17, 3*pi/2, 2*pi, mpf(10)**(-15))
print_result(f1, a[0], a[1])
b = bisection(f2, 17, 0.0001, pi/2, mpf(10)**(-15))
print_result(f2, b[0], b[1])
c = bisection(f3, 17, 1, 3, mpf(10)**(-15))
print_result(f3, c[0], c[1])

funkcja f1, miejsce zerowe: 4.730040744862704, f(x0): -8.7690271710627599e-16, liczba iteracji: 53
funkcja f2, miejsce zerowe: 0.8603335890193797, f(x0): 2.3765711620882257e-16, liczba iteracji: 51
funkcja f3, miejsce zerowe: 1.829383601933849, f(x0): 6.2450045135165055e-16, liczba iteracji: 51


Liczba iteracji potrzebnych do osiągnięcia dokładności $ 10^{-33} $

In [51]:
a = bisection(f1, 35, 3*pi/2, 2*pi, mpf(10)**(-33))
print_result(f1, a[0], a[1])
b = bisection(f2, 35, 0.0001, pi/2, mpf(10)**(-33))
print_result(f2, b[0], b[1])
c = bisection(f3, 35, 1, 3, mpf(10)**(-33))
print_result(f3, c[0], c[1])

funkcja f1, miejsce zerowe: 4.7300407448627040260240481008338848, f(x0): -8.5538372920636217379869020340588373e-34, liczba iteracji: 110
funkcja f2, miejsce zerowe: 0.86033358901937976248389342413766217, f(x0): 6.1238553700437890015139298643130111e-34, liczba iteracji: 106
funkcja f3, miejsce zerowe: 1.829383601933848817136212946814151, f(x0): 8.6666847497425613387519007416321729e-34, liczba iteracji: 110


### K pierwszych pierwiastków

Funkcja f1 składa się z cosinusa oraz cosinusa hiperbolicznego. Cosinus hiperboliczny jest zawsze dodatni, a cosinus jest funkcją okresową która przyjmuje również wartości ujemne. Łatwo więc będzie generować kolejne przedziały zawierające jeden pierwiastek oraz mające przeciwny znak funkcji na brzegach. 

## Zadanie 2 Metoda Newtona

In [52]:
def newton(f, f_d, precision, max_iter, start_point, error):
    mp.dps = precision
    x0 = mpf(start_point)
    x1 = mpf(x0 - (f(x0)/f_d(x0)))
    i = 1
    
    while abs(f(x1) - f(x0)) > error and i < max_iter:
        x0 = x1
        x1 = mpf(x0 - (f(x0)/f_d(x0)))   # root of the tangent line
        i += 1
    
    return [x0, i]

Liczba iteracji potrzebnych do osiągnięcia dokładności $ 10^{-7} $

In [53]:
a = newton(f1, f1_d, 12, 120, 3*pi/2, mpf(10)**(-7))
print_result(f1, a[0], a[1])
b = newton(f2, f2_d, 12, 120, 0.0001, mpf(10)**(-7))
print_result(f2, b[0], b[1])
c = newton(f3, f3_d, 12, 120, 1, mpf(10)**(-7))
print_result(f3, c[0], c[1])

funkcja f1, miejsce zerowe: 4.73004074486, f(x0): 1.06865627458e-11, liczba iteracji: 4
funkcja f2, miejsce zerowe: 0.860333589239, f(x0): -8.1286088971e-10, liczba iteracji: 17
funkcja f3, miejsce zerowe: 1.82938360251, f(x0): 2.37287167693e-9, liczba iteracji: 8


Liczba iteracji potrzebnych do osiągnięcia dokładności $ 10^{-15} $

In [54]:
a = newton(f1, f1_d, 17, 120, 3*pi/2, mpf(10)**(-15))
print_result(f1, a[0], a[1])
b = newton(f2, f2_d, 17, 120, 0.0001, mpf(10)**(-15))
print_result(f2, b[0], b[1])
c = newton(f3, f3_d, 17, 120, 1, mpf(10)**(-15))
print_result(f3, c[0], c[1])

funkcja f1, miejsce zerowe: 4.730040744862704, f(x0): -7.7195194680967916e-17, liczba iteracji: 5
funkcja f2, miejsce zerowe: 0.86033358901937976, f(x0): 0.0, liczba iteracji: 18
funkcja f3, miejsce zerowe: 1.8293836019338488, f(x0): 0.0, liczba iteracji: 9


Liczba iteracji potrzebna do osiągnięcia dokładności $ 10^{-33} $

In [55]:
a = newton(f1, f1_d, 35, 120, 3*pi/2, mpf(10)**(-33))
print_result(f1, a[0], a[1])
b = newton(f2, f2_d, 35, 120, 0.0001, mpf(10)**(-33))
print_result(f2, b[0], b[1])
c = newton(f3, f3_d, 35, 120, 1, mpf(10)**(-33))
print_result(f3, c[0], c[1])

funkcja f1, miejsce zerowe: 4.7300407448627040260240481008338848, f(x0): -1.6174802267314676109649814752178101e-34, liczba iteracji: 6
funkcja f2, miejsce zerowe: 0.86033358901937976248389342413766233, f(x0): -1.5046327690525280101999827676444745e-36, liczba iteracji: 19
funkcja f3, miejsce zerowe: 1.8293836019338488171362129468141508, f(x0): 0.0, liczba iteracji: 10


### Wnioski

Liczby iteracji potrzebne do osiągnięcia zadanych dokładnośc dla metody bisekcji nie różnią sie za bardzo od siebie między poszczególnymi funkcjami. Przy metodzie newtona wymagane jest dużo mniej iteracji niż przy metodzie bisekcji. Dokładne wartości dla metody newtona różnią sie jednak znacznie dla różnych funkcji.

## Zadanie 3 Metoda siecznych

In [56]:
def secant(f, precision, max_iter, a, b, error):
    mp.dps = precision
    x0 = mpf(a)
    x1 = mpf(b)
    x2 = mpf((f(x1) * x0 - f(x0) * x1)/(f(x1) - f(x0)))
    i = 0
    
    while abs(f(x2) - f(x1)) > error and i < max_iter:
        x0 = x1
        x1 = x2
        x2 = mpf((f(x1) * x0 - f(x0) * x1)/(f(x1) - f(x0)))
        i += 1
        
    return [x2, i]

Liczba iteracji potrzebna do osiągnięcia dokładności $ 10^{-7} $

In [57]:
a = secant(f1, 12, 120, 3*pi/2, 2*pi, mpf(10)**(-7))
print_result(f1, a[0], a[1])
b = secant(f2, 12, 120, 0.0001, pi/2, mpf(10)**(-7))
print_result(f2, b[0], b[1])
c = secant(f3, 12, 120, 1, 3, mpf(10)**(-7))
print_result(f3, c[0], c[1])

funkcja f1, miejsce zerowe: 4.73004074486, f(x0): 1.06865627458e-11, liczba iteracji: 5
funkcja f2, miejsce zerowe: 0.860333589019, f(x0): 2.27373675443e-13, liczba iteracji: 25
funkcja f3, miejsce zerowe: 1.82938360193, f(x0): 9.09494701773e-13, liczba iteracji: 9


Liczba iteracji potrzebna do osiągnięcia dokładności $ 10^{-15} $

In [58]:
a = secant(f1, 17, 120, 3*pi/2, 2*pi, mpf(10)**(-15))
print_result(f1, a[0], a[1])
b = secant(f2, 17, 120, 0.0001, pi/2, mpf(10)**(-15))
print_result(f2, b[0], b[1])
c = secant(f3, 17, 120, 1, 3, mpf(10)**(-15))
print_result(f3, c[0], c[1])

funkcja f1, miejsce zerowe: 4.730040744862704, f(x0): -7.7195194680967916e-17, liczba iteracji: 7
funkcja f2, miejsce zerowe: 0.86033358901937976, f(x0): 0.0, liczba iteracji: 26
funkcja f3, miejsce zerowe: 1.8293836019338488, f(x0): 6.9388939039072284e-18, liczba iteracji: 10


Liczba iteracji potrzebna do osiągnięcia dokładności $ 10^{-33} $

In [59]:
a = secant(f1, 35, 120, 3*pi/2, 2*pi, mpf(10)**(-33))
print_result(f1, a[0], a[1])
b = secant(f2, 35, 120, 0.0001, pi/2, mpf(10)**(-33))
print_result(f2, b[0], b[1])
c = secant(f3, 35, 120, 1, 3, mpf(10)**(-33))
print_result(f3, c[0], c[1])

funkcja f1, miejsce zerowe: 4.7300407448627040260240481008338848, f(x0): -1.6174802267314676109649814752178101e-34, liczba iteracji: 8
funkcja f2, miejsce zerowe: 0.86033358901937976248389342413766233, f(x0): -1.5046327690525280101999827676444745e-36, liczba iteracji: 28
funkcja f3, miejsce zerowe: 1.8293836019338488171362129468141508, f(x0): 0.0, liczba iteracji: 12


### Wnioski

Tabele podsumowywujące metody aproksymacji (liczba iteracji potrzebna do osiągnięcia konkretnej dokładności)

Metoda bisekcji

|        | $10^{-7}$     | $10^{-15}$    | $10^{-33}$ |
|:------:|:-------------:|:-------------:|:----------:|
| f1     | 28            | 53            | 111        |
| f2     | 20            | 51            | 106        |
| f3     | 22            | 51            | 110        |

Metoda newtona

|        | $10^{-7}$     | $10^{-15}$    | $10^{-33}$ |
|:------:|:-------------:|:-------------:|:----------:|
| f1     | 4             | 5             | 6          |
| f2     | 17            | 18            | 19         |
| f3     | 8             | 9             | 10         |

Metoda siecznych


|        | $10^{-7}$     | $10^{-15}$    | $10^{-33}$ |
|:------:|:-------------:|:-------------:|:----------:|
| f1     | 5             | 7             | 8          |
| f2     | 25            | 26            | 28         |
| f3     | 9             | 10            | 12


Ogólnie rzecz biorąc metody newtona oraz siecznych sprawują się dużo lepiej od bisekcji. Zaletą bisekcji jest to że nie ma parktycznie znaczenia jakiej funkcji przybliżamy pierwiastek. Przy pozostałych metodach widać znaczną różnice pomiędzy funkcjami. Metoda newtona wydaje się być nieco szybsza od metody siecznych, ale ta druga ma tą zalete że nie wymaga znajomości pochodnej funkcji, której pierwiastek przybliżamy. Pojęcia siecznej oraz stycznej są ze sobą mocno powiązane, dlatego też metody te dają zbliżone rezultaty.